src/clientHandler.ts

```diff
@@ -0,0 +1,252 @@
+import * as vscode from 'vscode';
+import ShortUniqueId from 'short-unique-id';
+import {
+	Executable,
+	LanguageClient,
+	LanguageClientOptions,
+	ServerOptions,
+	State,
+	DocumentSelector
+} from 'vscode-languageclient/node';
+import {
+	config,
+	getFolderName,
+	getWorkspaceFolder,
+	normalizeFolderName,
+	sortedWorkspaceFolders
+} from './vscodeUtils';
+import * as path from 'path';
+import TelemetryReporter from 'vscode-extension-telemetry';
+
+export interface TerraformLanguageClient {
+	commandPrefix: string,
+	client: LanguageClient
+}
+
+const clients: Map<string, TerraformLanguageClient> = new Map();
+
+/**
+ * ClientHandler maintains lifecycles of language clients
+ * based on the server's capabilities (whether multi-folder
+ * workspaces are supported).
+ */
+export class ClientHandler {
+	private shortUid: ShortUniqueId;
+	private pathToBinary: string;
+	private supportsMultiFolders = true;
+
+	constructor(private context: vscode.ExtensionContext, private reporter: TelemetryReporter ) {
+		this.shortUid = new ShortUniqueId();
+		this.pathToBinary = config('terraform').get('languageServer.pathToBinary');
+		if (this.pathToBinary) {
+			this.reporter.sendTelemetryEvent('usePathToBinary');
+		} else {
+			const installPath = path.join(context.extensionPath, 'lsp');
+			this.pathToBinary = path.join(installPath, 'terraform-ls');
+		}
+	}
+
+	public StartClients(folders?: string[]): vscode.Disposable[] {
```

Comment:
```suggestion
	public startClients(folders?: string[]): vscode.Disposable[] {
```
TypeScript style would be keeping a lowercase letter 🙂